### PR TITLE
dimod refresh part 2: BQMs, sampleset, decorators, asserts

### PR DESCRIPTION
**Pull request by JoelPasvolsky**

### CODE REVIEW

File: dimod/binary_quadratic_model.py [hunk 1]
```diff
@@ -120,12 +120,12 @@ class BinaryQuadraticModel(abc.Sized, abc.Container, abc.Iterable):
         >>> bqm = dimod.BinaryQuadraticModel({0: 1, 1: -1, 2: .5},
         ...                                  {(0, 1): .5, (1, 2): 1.5},
         ...                                  1.4,
-        ...                                  dimod.SPIN)
+        ...                                  dimod.Vartype.SPIN)
```
Comment (arcondello) [inline dimod/binary_quadratic_model.py]: why this change? SPIN and BINARY are in the top-level namespace

Comment (JoelPasvolsky) [reply dimod/binary_quadratic_model.py]: Yes, and I wanted to show that with the two consecutive examples using both supported formats rather than just the one

File: dimod/sampleset.py [hunk 2]
```diff
@@ -646,7 +650,22 @@ def first(self):
 
     @property
     def info(self):
-        """Dict of information about the :class:`SampleSet` as a whole."""
+        """Dict of information about the :class:`SampleSet` as a whole.
+
+        Examples:
+           This example shows the type of information that might be returned by
```
Comment (arcondello) [inline dimod/sampleset.py]: I am worried that this example is too specific. The sample set is used all over the place the info will not always be the same.

File: dimod/sampleset.py [hunk 3]
```diff
@@ -1178,16 +1218,17 @@ def lowest(self, rtol=1.e-5, atol=1.e-8):
                           self.vartype)
 
     def truncate(self, n, sorted_by='energy'):
-        """Create a new SampleSet with rows truncated after n.
+        """Create a new sample set with up to n rows.
```
Comment (arcondello) [inline dimod/sampleset.py]: I am worried that it is not clear that it takes the first n rows.

File: dimod/sampleset.py [hunk 4]
```diff
@@ -1257,23 +1298,23 @@ def slice(self, *slice_args, **kwargs):
             8  0  0  0  0  0  0  0  0  1  0      8       1
             9  0  0  0  0  0  0  0  0  0  1      9       1
             ['BINARY', 10 rows, 10 samples, 10 variables]
-
+            >>>
```
Comment (arcondello) [inline dimod/sampleset.py]: Why add the carrots?

File: dimod/utilities.py [hunk 5]
```diff
@@ -350,28 +350,35 @@ def resolve_label_conflict(mapping, old_labels=None, new_labels=None):
...
+        sampler found.
 
     Raises:
         ValueError: If no structured sampler is found.
 
+    Examples:
+
+    >>> import dimod
+    >>> sampler = dimod.StructureComposite(dimod.ExactSolver(),[0, 1], [(0, 1)])
+    >>> print(dimod.child_structure_dfs(sampler).nodelist)
```
Comment (arcondello) [inline dimod/utilities.py]: Probably we want this example to show the structure being nested. Maybe ```sampler = dimod.TrackingComposite(dimod.StructureComposite(dimod.ExactSolver(), [0, 1], [(0, 1)])))```

File: dimod/sampleset.py (hunk 2 above)
Comment (JoelPasvolsky) [reply dimod/sampleset.py]: The intro sentence is pretty strongly worded to that effect: "type of information that might be returned"

File: dimod/sampleset.py (hunk 3 above)
Comment (JoelPasvolsky) [reply dimod/sampleset.py]: I feel that with both (1) the function name, "truncate" and (2) the following description, "Does not return  any rows above this limit in the original" that the error probability is sufficiently low. If you are highly worried, I can change

File: dimod/sampleset.py (hunk 4 above)
Comment (JoelPasvolsky) [reply dimod/sampleset.py]: Keeps the whole thing in one example in RTD, with a single scroll bar, as opposed to partial examples, each with a scroll bar. (And doctest broke when I had ellipses.)

File: dimod/utilities.py (hunk 5 above)
Comment (JoelPasvolsky) [reply dimod/utilities.py]: Yes, thanks for this suggestion, I like it

File: dimod/sampleset.py (hunk 4 above)
Comment (arcondello) [reply dimod/sampleset.py]: Probably it would be better to just shorten the width then. E.g.
```
            >>> import numpy as np
            ...
            >>> sampleset = dimod.SampleSet.from_samples(np.diag(range(1, 11)),
            ...                                                                              dimod.BINARY, energy=range(10))
            >>> print(sampleset)
               0  1  2  3  4  5  6  7  8  9 energy num_oc.
            0  1  0  0  0  0  0  0  0  0  0      0       1
            1  0  1  0  0  0  0  0  0  0  0      1       1
            2  0  0  1  0  0  0  0  0  0  0      2       1
            3  0  0  0  1  0  0  0  0  0  0      3       1
            4  0  0  0  0  1  0  0  0  0  0      4       1
            5  0  0  0  0  0  1  0  0  0  0      5       1
            6  0  0  0  0  0  0  1  0  0  0      6       1
            7  0  0  0  0  0  0  0  1  0  0      7       1
            8  0  0  0  0  0  0  0  0  1  0      8       1
            9  0  0  0  0  0  0  0  0  0  1      9       1
            ['BINARY', 10 rows, 10 samples, 10 variables]
```

File: dimod/utilities.py (hunk 5 above)
Comment (arcondello) [reply dimod/utilities.py]: Even better on multiple lines
```
sampler = dimod.TrackingComposite(
    dimod.StructureComposite(
        dimod.ExactSolver(), [0, 1], [(0, 1)])))
```

Comment (JoelPasvolsky) [reply dimod/utilities.py]: Yes, I had changed to 
```
>>> import dimod
    >>> sampler = dimod.TrackingComposite(dimod.StructureComposite(dimod.ExactSolver(),
    ...                                   [0, 1], [(0, 1)]))
 >>>
```

Comment (arcondello) [reply dimod/utilities.py]: That still exceeds the 80 char width by a fair margin (and introduces scroll bars)

File: dimod/sampleset.py (hunk 4 above)
Comment (JoelPasvolsky) [reply dimod/sampleset.py]: That will get rid of the scrolls but will keep the separation into partial examples

File: dimod/utilities.py (hunk 5 above)
Comment (JoelPasvolsky) [reply dimod/utilities.py]: How about:
```
>>> sampler = dimod.TrackingComposite(dimod.StructureComposite(
...                 dimod.ExactSolver(), [0, 1], [(0, 1)]))
```

Comment (arcondello) [reply dimod/utilities.py]: I think I am attached to my example because it shows the 3-depth of the nesting clearly (I think). But your call.

Comment (JoelPasvolsky) [reply dimod/utilities.py]: Ah, I hadn't considered that aspect so I preferred two docstring lines to three but that is nice so will change.

File: dimod/sampleset.py (hunk 4 above)
Comment (arcondello) [reply dimod/sampleset.py]: I like the partial examples personally, but ymmv.

Comment (JoelPasvolsky) [reply dimod/sampleset.py]: Our other examples are self-contained so I'd go with consistency barring some improvement to clarity. 
"ymmv": my daily new acronym

Comment (randomir) [reply dimod/sampleset.py]: @JoelPasvolsky, I vote for no prompts between these examples. It's much nicer to have them self-contained. If you don't like the series of boxes on output, you can extract the comment line as a one line description of the example.

Comment (JoelPasvolsky) [reply dimod/sampleset.py]: Okay, looks like we have mob rule in sampleset. I split it without the comment lines:
image
